// @lijuhong1981/jsload/index.js
import { readAsArrayBuffer, readAsDataURL, readAsText } from "./src/FileRead.js";
import readLine from "./src/readLine.js";
import { inputFile, inputFileAsArrayBuffer, inputFileAsDataURL, inputFileAsText } from "./src/FileInput.js";
import Cache from "./src/Cache.js";
import loadImage from "./src/loadImage.js";
import Loader from "./src/Loader.js";
import ImageLoader from "./src/ImageLoader.js";
import arrayBufferToBinaryString from "./src/arrayBufferToBinaryString.js";
import arrayBufferToBinaryStringAsync from "./src/arrayBufferToBinaryStringAsync.js";
import blobToImage from "./src/blobToImage.js";
import blobToImageAsync from "./src/blobToImageAsync.js";
import arrayBufferToImage from "./src/arrayBufferToImage.js";
import arrayBufferToImageAsync from "./src/arrayBufferToImageAsync.js";
import fetchResponse from "./src/fetchResponse.js";
import fetchArrayBuffer from "./src/fetchArrayBuffer.js";
import fetchBlob from "./src/fetchBlob.js";
import fetchJson from "./src/fetchJson.js";
import fetchText from "./src/fetchText.js";
import fetchBinaryString from "./src/fetchBinaryString.js";
import fetchImageBitmap from "./src/fetchImageBitmap.js";

export {
    readAsArrayBuffer,
    readAsDataURL,
    readAsText,
    readLine,
    inputFile,
    inputFileAsArrayBuffer,
    inputFileAsDataURL,
    inputFileAsText,
    Cache,
    loadImage,
    Loader,
    ImageLoader,
    arrayBufferToBinaryString,
    arrayBufferToBinaryStringAsync,
    blobToImage,
    blobToImageAsync,
    arrayBufferToImage,
    arrayBufferToImageAsync,
    fetchResponse,
    fetchArrayBuffer,
    fetchBlob,
    fetchJson,
    fetchText,
    fetchBinaryString,
    fetchImageBitmap,
};
